### ISSUE TITLE
Consider defining a helper "health check" utility for distroless scenarios

### ISSUE THREAD

**Issue by mthalman** [dotnet/dotnet-docker#4300]
The use of the `HEALTHCHECK` instruction often involves the use of a command that executes an HTTP request to the application in order to verify its health. That command might be `curl` or `wget` for Linux containers. This becomes problematic for distroless/chiseled containers that would likely want to exclude such commands. How then would a call be made to verify the health of the app?

One solution is to include a basic .NET application that execute the HTTP request (see https://github.com/dotnet/dotnet-docker/discussions/4296). Depending on the ubiquity of such a solution, it might make sense for .NET to provide its own implementation of this application that can be easily injected into .NET distroless/chiseled solutions.

I'll illustrate this with an example. Let's say .NET publishes an `httpservice` image that contains a .NET app which simply sends an HTTP request to a URL provided as an argument. It returns 0 if the request succeeds and 1 if it fails. A distroless implementation could consume this app like the following:

```Dockerfile
# Build stage excluded for brevity
...

FROM mcr.microsoft.com/dotnet/nightly/runtime-deps:7.0-jammy-chiseled
WORKDIR /app
COPY --from=build /app .

# Inject the .NET httpservice into the image to use for the health check
COPY --from=mcr.microsoft.com/dotnet/httpservice:7.0 /httpservice.exe /
HEALTHCHECK CMD [ "/httpservice.exe", "http://localhost/healthz" ]

ENTRYPOINT ["./app"]
```

I think it'd be important to understand the landscape of how customers are using health checks to determine the suitability of a pre-defined tool like this. We obviously don't want to be reimplementing curl in .NET but allowing for a little bit of behavior customization in order to fulfill a large portion of customer scenarios seems reasonable.

**Comment by mthalman** [comment]
[Triage]
As a first step, we'd like to provide documentation in the repo to help guide customer's on the pattern to implement this on their own in a way that doesn't depend on .NET providing a new tool. This is a low-cost solution and can also be used as a means to gather feedback from customers to determine the optimal solution that would meet their needs if we were to publish our own tool.

**Comment by 64J0** [comment]
Does this affect Kubernetes' probes? 

I'm not 100% sure, but I think the probes commands are handled outside the container, so it must work even in a distroless environment. Is it correct?

**Comment by richlander** [comment]
> Does this affect Kubernetes' probes?

Not needed for Kubernetes. This is a docker compose need.

**Comment by Kralizek** [comment]
As far as I understand, having a way to ping the application from within the container is very important in the AWS space for people (like me) using ECS instead of EKS.

**Comment by mu88** [comment]
@mthalman: is there any news about this? I have a bunch of distroless ASP.NET Core apps running on my Raspi and I'd like to make use of a simple health check probe to http://localhost:8080/healthz in my `docker-compose.yml`. Or is there any prototype that I can give a shot so that you get some early feedback and I have a MVP? 🙂

**Comment by mthalman** [comment]
> is there any news about this?

There's been recent discussion on this as part of .NET 10 planning. Nothing is decided yet. But we recognize the interest in this feature.

**Comment by mu88** [comment]
But for the moment, there is no alternative/hack, correct?

**Comment by MichelZ** [comment]
We have a mini-healthcheck binary that we include in the container:
```

 static async Task<int> Main(string[] args)
 {
     var httpClient = new HttpClient();
     httpClient.DefaultRequestHeaders.ConnectionClose = true;

     if (args.Length > 1 && Uri.TryCreate(args[1], UriKind.Absolute, out var uri))
         return await httpClient.GetAsync(uri).ContinueWith(r => r.Result.IsSuccessStatusCode) ? 0 : 1;
     else
         throw new ArgumentNullException("A valid URI must be given as first argument");
 }
```

DOCKERFILE:

```
# Build the HealthChecker
WORKDIR "/src/src/container.HealthChecker"
RUN dotnet publish -c Release --runtime linux-x64 --no-self-contained "container.HealthChecker.csproj" -o "/app/publish/HealthChecker" 

[...]

HEALTHCHECK CMD [ "dotnet", "/app/HealthChecker/container.HealthChecker.dll", "--", "http://localhost:8080/health" ]
```

**Comment by mu88** [comment]
Thx for this hint, maybe that's a viable workaround for me 👍🏻 Since I'm using the [.NET SDK Container Building Tools](https://learn.microsoft.com/en-us/dotnet/core/docker/publish-as-container), I'd have to build an intermediary Docker base image (e.g. `mu88/dotnet-aspnet:9.0`) as my apps do no longer have a `Dockerfile`.

@mthalman: when discussing this particular feature internally, maybe it's worth taking this container publishing approach without an actual `Dockerfile` into account, too. For example, I could imagine that MS publishes slightly extended base images (e.g. `dotnet/aspnet:9.0-chiseled-health`) which contain both the runtime (here ASP.NET Core chiseled) and the "mini health check tool". This way, consumers using the .NET SDK Container Building Tools could simply use this as the base image without building it alone.

CC @baronfel

**Comment by Kralizek** [comment]
I agree wholeheartedly with @mu88 . I also use the `dotnet publish` approach and I would hate having to return to the dockerfile approach to leverage this feature.

TBF/OT, the SDK approach needs to be extended to extra customization regardless of this or that approach.

**Comment by mthalman** [comment]
@baronfel - Does the SDK container publishing even have a means to define `HEALTHCHECK`, regardless of whether distroless is being used?

**Comment by baronfel** [comment]
We do not support HEALTHCHECK in any way in the SDK tooling (our issue discussing it was https://github.com/dotnet/sdk-container-builds/issues/316). We chose not to specifically because OCI images (where the world is moving to) doesn't make use of them.

**Comment by mu88** [comment]
MichelZ's approach inspired me. However, I decided against my first idea and didn't deploy my little health check tool via my Docker base image. Since I use several different .NET base images (`dotnet/aspnet:9`, `dotnet/aspnet:9-noble-chiseled`, `dotnet/aspnet:9-noble-chiseled-extra`), I'd have to maintain several different custom base images, too. As I'm lazy and don't like `Dockerfile`s 😆 I ship it as part of my NuGet package `mu88.Shared` as it already contains some shared logic for all my .NET apps. [Here you find the corresponding commit](https://github.com/mu88/mu88.Shared/commit/56e7c14ba58a85c8b7eb1d951b28cd21dcb741f3).

It took me a while to deploy both the health check assembly and the necessary `runtimeconfig.json` properly via NuGet so that these files flow through to the app's output directory 😮‍💨 as I ship this file as a `<Content>` item, the `mu88.HealthCheck.runtimeconfig.json` now is displayed in the IDE's _Solution Explorer_:  
![Image](https://github.com/user-attachments/assets/3b223fab-c756-4165-a97b-659a434a8aa6)


That's not really nice, but so far I haven't found a solution for this (setting it to `<None>` doesn't work) - maybe someone here has a deeper understanding of the underlying NuGet workflows and/or a good idea 🙂

Anyhow, now it works properly with this `docker-compose.yml` and I can still use the .NET SDK Container Building Tools 🤓🥳
```yaml
services:
  myservice:
    container_name: mycontainer
    image: mu88/myimage:latest-chiseled
    healthcheck:
      test: [ "CMD", "dotnet", "/app/mu88.HealthCheck.dll", "http://localhost:8080/healthz" ]
      interval: 60s
      timeout: 5s
      retries: 3
      start_period: 30s
      start_interval: 5s
```

If the community should be interested, I could also ship this component as a dedicated NuGet package. But for the moment, I keep it in my already existing package as I don't expect much interest in this niche feature.

Thx again to @MichelZ for your input 🙏🏻

**Comment by baronfel** [comment]
Glad you found a way to make things work! I think for cross-cutting tools like this the answer that other container platforms have settled on is making specific container layers with apps/diagnostic tooling/etc and then having the container publishing process slip-stream in those layers.

Since a layer is just a file system tarball, practically speaking this means something like:
* publishing the app/tool
* placing it in a file system layout
* creating a tarball of this layout
* pushing the tarball to a registry to create a referenceable Layer
* instructing container building tools to include this Layer

For arch-specific tools this would mean keeping a (signed and verified) inventory of per-RID Layer digests, etc. The SDK Container tooling would also need to implement https://github.com/dotnet/sdk-container-builds/issues/325 to allow pointing to arbitrary Layers.  So there's a decent amount of pre-work to get to this final state.

**Comment by alexaka1** [comment]
I was also bothered on having to give up on healthchecks when using the chiseled images, so I made a docker image that contains a small AOT app, that can perform healthchecks.
https://hub.docker.com/repository/docker/alexaka1/distroless-dotnet-healthchecks

To overcome CPU architectures, I utilized docker's cross compilation, and I have tested it on a macbook (I don't have access to ARM github runners 😢). It works with any .NET 8+ image (even the minimum runtime-deps:*-chiseled-aot image) and it is zero config out of the box.
